src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,434 @@
+            result_type = std::make_shared<DataTypeArray>(std::move(result_type));
+    }
+    return result_type;
+}
+
+/*
+Checks, in what endian format data was written.
+return -1: if data is written in little-endian;
+
+         1: if data is written in big-endian;
+
+         0: if data is written in no-endian. */
+int endianOrientation(String descr)
+{
+    if (descr.length() < 3)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Descr field length must be bigger or equal 3");
+    if (descr[0] == '<')
+        return -1;
+    else if (descr[0] == '>')
+        return 1;
+    else if (descr[0] == '|')
+        return 0;
+    else
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Wrong content of field descr");
+}
+
+DataTypePtr parseType(String type)
+{
+    if (type == "<i1")
+        return std::make_shared<DataTypeInt8>();
+    else if (type == "<i2")
+        return std::make_shared<DataTypeInt16>();
+    else if (type == "<i4")
+        return std::make_shared<DataTypeInt32>();
+    else if (type == "<i8")
+        return std::make_shared<DataTypeInt64>();
+    else if (type == "<u1")
+        return std::make_shared<DataTypeUInt8>();
+    else if (type == "<u2")
+        return std::make_shared<DataTypeUInt16>();
+    else if (type == "<u4")
+        return std::make_shared<DataTypeUInt32>();
+    else if (type == "<u8")
+        return std::make_shared<DataTypeUInt64>();
+    else if (type == "<f2")
+        return std::make_shared<DataTypeFloat32>();
+    else if (type == "<f4")
+        return std::make_shared<DataTypeFloat32>();
+    else if (type == "<f8")
+        return std::make_shared<DataTypeFloat64>();
+    else if (type == "<c8" || type == "<c16")
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support complex numeric type");
+    else if (type == "|b1")
+        return std::make_shared<DataTypeInt8>();
+    else if (type[1] == 'U' || type[1] == 'S')
+        return std::make_shared<DataTypeString>();
+    else if (type == "O")
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support object types");
+    else
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Error while parsing data type");
+}
+
+std::vector<int> parseShape(String shapeString)
+{
+    shapeString.erase(std::remove(shapeString.begin(), shapeString.end(), '('), shapeString.end());
+    shapeString.erase(std::remove(shapeString.begin(), shapeString.end(), ')'), shapeString.end());
+
+    // Use a string stream to extract integers
+    String value;
+
+    std::vector<int> shape;
+
+    size_t start = 0, end = 0;
+
+    while ((end = shapeString.find(',', start)) != std::string::npos)
+    {
+        shape.push_back(std::stoi(shapeString.substr(start, end - start)));
+        start = end + 1;
+    }
+
+    // Add the last token (or the only token if no delimiter is found)
+    if (start != shapeString.length())
+        shape.push_back(std::stoi(shapeString.substr(start)));
+    return shape;
```

Comment:
Maybe we can do it in a more simple way splitting string by comma using `boost::split` and applying `std::stoi` to splitted strings. I think we can do it using fewer lines of code